### DISCUSSION TITLE
Unstructured Grid to binary array

### DISCUSSION THREAD

**Discussion by koenterheegde0507** [pyvista/pyvista#6042] [Q&A]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [X] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

```python
from skimage.morphology import skeletonize 
final_mesh = grid.extract_surface()

density = final_mesh.length / 100
x_min, x_max, y_min, y_max, z_min, z_max = final_mesh.bounds
x = np.arange(x_min, x_max, density)
y = np.arange(y_min, y_max, density)
z = np.arange(z_min, z_max, density)
x, y, z = np.meshgrid(x, y, z)

# Create unstructured grid from the structured grid
structured_grid = pv.StructuredGrid(x, y, z)
ugrid = pv.UnstructuredGrid(structured_grid)

# get part of the mesh within the mesh's bounding surface.
selection = ugrid.select_enclosed_points(mesh.extract_surface(),
                                         tolerance=0.0,
                                         check_surface=False)
mask = selection.point_data['SelectedPoints'].view(bool)
mask = mask.reshape(x.shape)

pv.plot(structured_grid.points, scalars=mask)

skeleton = skeletonize(mask)
```

### Description

I'm currently struggling to obtain a binary array from either my mesh or an unstructured grid. Context: I generate a polydata from a marching cubes algorithm, based on a binary volume as input from a CT scan segmentation. I then want to edit this mesh in 3D and therefore have to cast it to an unstructured grid. Once I'm done editing the grid, I go back to a mesh again, and that's where this code starts. I generate the mesh using `final_mesh = grid.extract_surface()`. The reason I do this is because I saw this issue: https://github.com/pyvista/pyvista-support/issues/141, in which they also start with a mesh and I can't find any solution on how to go from an unstructured grid to binary array immediately. Anyway, I use the code that was used in this issue, but it doesn't seem to work in my case. 
This is my unstructured grid: 
![grid_issue](https://github.com/pyvista/pyvista/assets/152174547/269df536-5226-4bbe-8072-323938413547)
And this is binary mask:
![binary_array](https://github.com/pyvista/pyvista/assets/152174547/d974561b-5ec7-4f6b-8f10-cbb1db8957cd)

Anyone knows what I'm doing wrong or has any other approach? 
Appreciate the help!

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Sun May 05 20:54:54 2024 West-Europa (zomertijd)

                OS : Windows
            CPU(s) : 4
           Machine : AMD64
      Architecture : 64bit
       Environment : Python
        GPU Vendor : ATI Technologies Inc.
      GPU Renderer : Radeon (TM) RX 480 Graphics
       GPU Version : 4.6.0 Compatibility Profile Context 24.1.1.231127
  MathText Support : False

  Python 3.11.0 (main, Oct 24 2022, 18:26:48) [MSC v.1933 64 bit (AMD64)]

           pyvista : 0.43.5
               vtk : 9.3.0
             numpy : 1.26.4
        matplotlib : 3.8.4
            scooby : 0.9.2
             pooch : 1.8.1
            pillow : 10.3.0
           imageio : 2.34.0
             PyQt5 : 5.15.10
             scipy : 1.13.0
            meshio : 5.3.5
--------------------------------------------------------------------------------
```

**Comment by user27182** [comment]
This isn't a full solution but you could try:

1. use [voxelize](https://docs.pyvista.org/version/stable/api/utilities/_autosummary/pyvista.voxelize.html#pyvista-voxelize) to "regularize" the points in your final, processed mesh and make them uniform.
2. then, use [cell_centers](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.datasetfilters.cell_centers#pyvista.DataSetFilters.cell_centers) and use the center of the voxel cells as a representation of the `True` points from your binary array. 
3. then, perhaps do a search using something like [find_closest_point](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.dataset.find_closest_point) to find which `True` points of your unstructured grid (i.e. the cell centres of the voxelized mesh) correspond to the points in the original input image (i.e. the one you originally applied marching cubes to)

**Comment by koenterheegde0507** [comment]
I found out that I have another issue: The polydata I generate from the marching cubes algorithm generates a surface instead of a volume. I do need the whole volume as a binary array for the rest of my code to work. So I either have to find a way to visualize and edit this volume in 3D, so I don't need to use marching cubes anymore or I have to find a way to go from a surface mesh to a volumetric mesh and then do the steps that you gave. Do you have any idea how to solve this?

**Reply by koenterheegde0507** [reply]
I found the answer! There were some holes in the mesh which made the other code work incorrectly. So it has nothing to do with the issues I mentioned.
